src/ir/bits.h

```diff
@@ -128,35 +128,100 @@ struct DummyLocalInfoProvider {
 template<typename LocalInfoProvider = DummyLocalInfoProvider>
 Index getMaxBits(Expression* curr,
                  LocalInfoProvider* localInfoProvider = nullptr) {
-  if (auto* const_ = curr->dynCast<Const>()) {
+  if (auto* c = curr->dynCast<Const>()) {
     switch (curr->type.getSingle()) {
       case Type::i32:
-        return 32 - const_->value.countLeadingZeroes().geti32();
+        return 32 - c->value.countLeadingZeroes().geti32();
       case Type::i64:
-        return 64 - const_->value.countLeadingZeroes().geti64();
+        return 64 - c->value.countLeadingZeroes().geti64();
       default:
         WASM_UNREACHABLE("invalid type");
     }
   } else if (auto* binary = curr->dynCast<Binary>()) {
     switch (binary->op) {
       // 32-bit
-      case AddInt32:
-      case SubInt32:
-      case MulInt32:
-      case DivSInt32:
-      case DivUInt32:
-      case RemSInt32:
-      case RemUInt32:
       case RotLInt32:
       case RotRInt32:
+      case AddInt32:
+      case SubInt32:
+        return 32;
+      case MulInt32: {
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        if (maxBitsRight == 32) {
+          return 32;
```

Comment:
this is not needed - the return lower down handles it already.

Reply:
Yes, but it potentially could speedup this process. The probability that the value will be negative is very high. Or I miss something?

Reply:
I don't think it's worth trying to speed up the pass at the cost of making the code redundant and complex.

If we find big slowdowns in profiling, we can consider optimization work. But otherwise I think simplicity is the much more important factor.

Reply:
Done!